### PR TITLE
Support for Features with Different Dimensions in NormalizedMI

### PR DESCRIPTION
**Pull request by marko-tuononen**
This PR introduces support for the NormalizedMI class to handle features with varying dimensions. Previously, all features were required to have the same dimension, but this update allows for both single integer n_dims and lists of integers representing different feature dimensions. The implementation includes updates to _check_X() and other relevant methods. Unit tests have been added and modified to verify the functionality. This enhancement improves the flexibility and applicability of the NormalizedMI estimator.

### CODE REVIEW

**Comment by marko-tuononen** [comment]
@branii Hi Daniel, I wanted to bring this PR to your attention. It introduces support for features with different dimensions in the NormalizedMI class. I'd appreciate your review when you have a moment. Thank you!

**Comment by braniii** [comment]
@marko-tuononen Thank you so much for contributing to this package. I have never thought about this scenario, but it seems to be a useful enhancement.

File: src/normi/_estimators.py [hunk 1]
```diff
@@ -94,15 +94,15 @@ class NormalizedMI(BaseEstimator):
     def __init__(
         self,
         *,
-        n_dims: PositiveInt = 1,
+        n_dims: Union[List[PositiveInt], PositiveInt] = 1,
```
Comment (braniii) [inline src/normi/_estimators.py:97]: Instead of using `List[PositiveInt]` i guess it would be helpful to accept numpy arrays as well. This could be achieved by introducing a `ArrayLikeInt` and `ArrayLikePositiveInt` similar to https://github.com/moldyn/NorMI/blob/main/src/normi/_typing.py#L87-L88

File: src/normi/_estimators.py [hunk 2]
```diff
@@ -94,15 +94,15 @@ class NormalizedMI(BaseEstimator):
...
         normalize_method: NormString = 'geometric',
         invariant_measure: InvMeasureString = 'volume',
         k: PositiveInt = 5,
         n_jobs: Int = -1,
         verbose: bool = True,
     ):
         """Initialize NormalizedMI class."""
-        self.n_dims: PositiveInt = n_dims
+        self.n_dims: list[PositiveInt] = n_dims
```
Comment (braniii) [inline src/normi/_estimators.py:105]:
```suggestion
        self.n_dims: List[PositiveInt] = n_dims
```

Comment (braniii) [reply src/normi/_estimators.py:105]: `self.n_dims` should have the same type as `n_dims`. But I guess this should become something like `Union[PositiveInt, ArrayLikePostiveInt]`.

File: src/normi/_estimators.py [hunk 3]
```diff
@@ -131,21 +131,26 @@ def fit(
...
         # scale input
         X = StandardScaler().fit_transform(X)
-        X = np.split(X, self._n_features, axis=1)
+        if isinstance(self.n_dims, int):
+            X = np.split(X, self._n_features, axis=1)
+        else:
+            X = np.split(X, np.cumsum(self.n_dims), axis=1)[:-1]
 
-        self.mi_: PositiveMatrix
+        self.mi_: FloatMatrix
```
Comment (braniii) [inline src/normi/_estimators.py:153]: Did you face an issue similar to #6 or #7? In theory, the mutual information should be positive. But I guess the same holds for the entropy, so this would solve raising ugly beartype errors.

File: tests/test__estimators.py [hunk 4]
```diff
@@ -77,6 +77,10 @@ def test__scale_nearest_neighbor_distance(
 @pytest.mark.parametrize('X, n_dims, error', [
     (np.random.uniform(size=(10, 9)), 1, None),
     (np.random.uniform(size=(10, 9)), 3, None),
+    (np.random.uniform(size=(10, 9)), [3, 3, 3], None),
```
Comment (braniii) [inline tests/test__estimators.py:80]: I guess it would be great to add one passing test for a new split, e.g.` [4, 3, 2]`. And to check for a passing numpy array, e.g. `np.array([4, 3,2])`

**Review by braniii — COMMENTED** [review]
This PR looks really good. I've added some minor comments.

File: src/normi/_estimators.py (hunk 1 above)
Comment (marko-tuononen) [reply src/normi/_estimators.py:97]: Numpy arrays are now accepted.

File: src/normi/_estimators.py (hunk 2 above)
Comment (marko-tuononen) [reply src/normi/_estimators.py:105]: Good catch. This is fixed.

File: src/normi/_estimators.py (hunk 3 above)
Comment (marko-tuononen) [reply src/normi/_estimators.py:153]: Yes. I am suffering negative MIs and entropies. This enables to get _some results even when there are some numerical problems leading to negative values.

**Comment by braniii** [comment]
@marko-tuononen sorry for the late answer. Thx for contributing to NorMI, the PR looks all good now :)

**Comment by marko-tuononen** [comment]
> @marko-tuononen sorry for the late answer. Thx for contributing to NorMI, the PR looks all good now :)

No worries, @braniii! I’m glad to have contributed to NorMI. Actually, I have another fix in the works and just submitted a paper related to it. I hope to find time soon to create a merge request for that as well.
